### PR TITLE
Enable saved payments from setup intents for payment methods that are converted to SEPA

### LINKED ISSUE #2010: UPE: Save non-card payment methods from a setup intent
As we were experimenting with UPE and different combinations of payment methods and paying for free trial subscriptions, it was discovered that we were unable to pay for a free trial subscription using Bancontact. This was alleviated by marking Bancontact as not reusable which was not a problem as this matched the original gateway's functionality. Ultimately it was determined that the issue with that combination was due to the inability to [save a payment method generated from a setup intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/78154394ec1b49c859c0ac949a25bdacb21a9ac5/includes/payment-methods/class-wc-stripe-upe-payment-method.php#L281-L283). This is needed by the following line to find the `generated_sepa_debit` to be stored as the token.

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/78154394ec1b49c859c0ac949a25bdacb21a9ac5/includes/payment-methods/class-wc-stripe-upe-payment-method.php#L223

I believe there should be some way to save these payment methods to a subscription so they could be charged in the future. Some inspiration on this could be found in #1977 as this leverages the saved payment methods being stored on Stripe. Some further context on the discssion can be found here: p1632862805343600-slack-C02BHNRN6R1

#### Steps to reproduce the issue

- First mark Bancontact as reusable in UPE: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/78154394ec1b49c859c0ac949a25bdacb21a9ac5/includes/payment-methods/class-wc-stripe-upe-payment-method-bancontact.php#L22
- Enable UPE and Bancontact within UPE.
- Set up WooCommerce Subscriptions with a subscription product with a free trial.
- Go to checkout with that product and use Bancontact to make the payment.
- An error should be thrown as the plugin is unable to save the payment method from the setup intent.

### LINKED ISSUE #2035: `create_payment_token_for_user()` visibility and parameters mismatch.
In [`class-wc-stripe-upe-payment-method-cc.php`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php#L112-L124) it's `public`, but in [`class-wc-stripe-upe-payment-method.php`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/payment-methods/class-wc-stripe-upe-payment-method.php#L257-L266) it's `private`.

I had a quick discussion with @superdav42 and he suggested that they should both be private and that their parameters should also be the same.

@FangedParakeet cc-ing you since you had opened [the issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2014) for adding testing for some of these methods.

### PR DESCRIPTION
**Pull request by FangedParakeet**
# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #2010 
Fixes #2035 

Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.

Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.

This PR fixes the saved payment flow with setup intents for payment methods that are converted to SEPA Direct Debits. When saving one of these payment methods using a payment intent, we need to scour the payment method details within the charge on the intent to find the generated SEPA payment method ID that we would like to store and reuse. Setup intents do not have charges, but instead there is a secret field called `latest_attempt` that we can expand to find the corresponding generated payment method details. 

This PR also changes the saved payment flow in a more general manner. Previously, we created a WC Payment Token and stored it during the checkout: that is actually not necessary, as whenever we query for WC Payment Tokens, we additionally query Stripe to see if new payment methods have been added and create and store tokens for these new payment methods on the fly. To support this, I've removed the function call to create a token at checkout and now we simply clear the customer's cache, to ensure that we make a new request to query Stripe for payment methods, rather than using the saved transient.

This change also allowed me to significantly thin out and refactor the UPE payment method class. I've added unit test coverage for all the flows in this PR and update the existing tests to play along as well. Additionally, while #2010 only specifically mentions Bancontact, I have also set the iDEAL and Sofort payment methods to be reusable as well, as they should work the same way. If you think I might be acting too big for my boots, please let me know and I'll yield and switch this property for these other payment methods back to `false`.

# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Enable UPE and Bancontact within UPE.
2. Set up WooCommerce Subscriptions with a subscription product with a free trial.
3. Go to checkout with that product and use Bancontact to make the payment.
4. Authorize the transaction. 
5. Order should complete successfully and the Order Received page should show the correct payment method. 
6. Select the Subscription (either from the Order Received page or your Accounts page) and confirm that the Bancontact payment method is listed as the payment method for the subscription.
7. Ensure that this payment method is saved as a SEPA Direct Debit in your Account's list of saved payment methods.
8. Ensure that you can complete a checkout with this saved payment method.

Other things that are perhaps worth testing as well
 - Purchasing a regular subscription product (without a free-trial and with a payment intent) with Bancontact.
 - Purchasing a regular product and saving the payment method with Bancontact.
 - All of the above with iDEAL and SOFORT.
 - All existing saved payment method flows (sorry).
-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: includes/class-wc-stripe-customer.php [hunk 1]
```diff
@@ -723,20 +723,4 @@ public static function map_customer_data( WC_Order $wc_order = null, WC_Customer
...
-	 * @param object $payment_method Payment method to be added.
-	 *
-	 * @since 5.6.0
-	 * @version 5.6.0
-	 */
-	public function add_payment_method_actions( $token, $payment_method ) {
-		// Clear cached payment methods.
-		$this->clear_cache();
-		do_action( 'woocommerce_stripe_add_payment_method', $this->get_id(), $token, $payment_method );
-	}
```
Comment (FangedParakeet) [inline includes/class-wc-stripe-customer.php:741]: I originally spawned this function, never liked it, and now I've finally done right by myself and put my son to rest for good. Good riddance.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 2]
```diff
@@ -861,13 +869,11 @@ public function process_order_for_confirmed_intent( $order, $intent_id, $save_pa
...
 	 *
 	 * @param object           $payment_method Stripe payment method object response.
-	 * @param WC_Payment_Token $token          WC Payment Token.
 	 *
 	 * @return object
 	 */
-	public function prepare_payment_method( $payment_method, $token ) {
+	public function prepare_payment_method( $payment_method ) {
 		return (object) [
-			'token_id'              => $token->get_id(),
```
Comment (FangedParakeet) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:870]: I don't think this property ever gets used anywhere in any UPE flow. I think it was just added to match the corresponding Sources flow, where it does get used.

File: includes/payment-methods/class-wc-stripe-upe-payment-method-ideal.php [hunk 3]
```diff
@@ -19,7 +19,7 @@ public function __construct() {
 		parent::__construct();
 		$this->stripe_id            = self::STRIPE_ID;
 		$this->title                = __( 'Pay with iDEAL', 'woocommerce-gateway-stripe' );
-		$this->is_reusable          = false; // TODO: Enable it in a future release - Supported through Stripe via SEPA Direct Debit.
+		$this->is_reusable          = true;
```
Comment (FangedParakeet) [inline includes/payment-methods/class-wc-stripe-upe-payment-method-ideal.php:22]: Yay or nay?

File: includes/payment-methods/class-wc-stripe-upe-payment-method-sofort.php [hunk 4]
```diff
@@ -19,7 +19,7 @@ public function __construct() {
 		parent::__construct();
 		$this->stripe_id            = self::STRIPE_ID;
 		$this->title                = __( 'Pay with SOFORT', 'woocommerce-gateway-stripe' );
-		$this->is_reusable          = false; // TODO: Enable it in a future release - Supported through Stripe via SEPA Direct Debit.
+		$this->is_reusable          = true;
```
Comment (FangedParakeet) [inline includes/payment-methods/class-wc-stripe-upe-payment-method-sofort.php:22]: Also, yay or nay?

File: tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php [hunk 5]
```diff
@@ -336,18 +347,20 @@ public function test_checkout_without_payment_uses_setup_intents() {
...
-			->with( "setup_intents/$setup_intent_id?expand[]=payment_method" )
+			->with( "setup_intents/$setup_intent_id?expand[]=payment_method&expand[]=latest_attempt" )
 			->will(
 				$this->returnValue(
 					$this->array_to_object( $setup_intent_mock )
 				)
 			);
 
-		$this->mock_gateway->process_upe_redirect_payment( $order_id, $setup_intent_id, false );
+		$this->mock_gateway->process_upe_redirect_payment( $order_id, $setup_intent_id, true );
```
Comment (FangedParakeet) [inline tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php:364]: I'm pretty sure if we're using a Setup Intent, by definition we are also saving the payment method.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 6]
```diff
@@ -1165,7 +1171,10 @@ private function get_payment_method_data_from_intent( $intent ) {
 				$payment_method_type    = ! empty( $payment_method_details ) ? $payment_method_details['type'] : '';
 			}
 		} elseif ( 'setup_intent' === $intent->object ) {
-			if ( ! empty( $intent->payment_method ) ) {
+			if ( ! empty( $intent->latest_attempt ) ) {
+				$payment_method_details = (array) $intent->latest_attempt->payment_method_details;
+				$payment_method_type    = ! empty( $payment_method_details ) ? $payment_method_details['type'] : '';
```
Comment (superdav42) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Looks like maybe we should pall back to using $intent->payment_method if $intent->latest_attempt->payment_method_details is empty. Like:
```suggestion
			if ( ! empty( $intent->latest_attempt ) && ! empty( $intent->latest_attempt->payment_method_details ) ) {
				$payment_method_details = (array) $intent->latest_attempt->payment_method_details;
				$payment_method_type    = $payment_method_details['type'];
```

**Review by superdav42 — APPROVED** [review]
Code looks good. Work for saving bancontact, iDeal or Sofort on checkout or in payment methods page. Saved payment methods work too. Tests makes sense and work.
One small suggestion.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 6 above)
Comment (FangedParakeet) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Nice catch, I completely agree.

**Pull request merged**